Script for building SingleR model

allyhawkins opened the pull request:
Partially addresses #270 

Here I started to add in a step to the `scpca-nf` workflow for building and training the models for the reference datasets to use when classifying cell types with `SingleR`. I am just starting by adding in a script that will take as input the reference data, stored as a `SummarizedExperiment` in a rds file and output a rds file with a list of two models, one containing a model using the `label.fine` and one containing `label.main`. This script will then be executed within a process in the `scpca-nf` workflow for creating the models, where all references for the specified project(s) or samples from a project being processed will go through the training step. 

A few comments on decisions I made here: 

- Right now, and for the forseeable future we are only using the `celldex` references, so because of that I didn't make an option to specify which labels to build the model with. I am using both labels each time. This will not work if the reference is not `celldex`, because the labels will be different. One approach is to make the labels to use an argument. Rather than make an argument, I chose to make models with both types of labels for now for testing purposes, but if others think we should change it I'm open to trying it. 
- The other thing I had to incorporate here was reading in the tx2gene reference file we use during the mapping step. The reason for that is I think we want to provide the genes to restrict when creating the model, and all of the SCE objects will contain the genes in this file. Alternatively we could just not restrict the genes? 

Once we have this script in, we can then add a new process to the workflow to run this script on each reference file. We will also need to create a metadata file that has the project id's and references for each project before we do that. I started to play around with this, but it was getting kind of tricky so I wanted to get some feedback just on the building the model part first before getting to the nextflow part. The other thought I had was for development purposes to keep the cell type module separate right now so testing and development is much easier and faster. Then once we feel confident about that we can write the code to link `main.nf` to the cell type module.

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,106 @@
+
+# combine into one object 
+all_models <- list(fine = fine_model,
+     main = main_model)
```
I think spacing is funky here? I might expect `main` to have ended up right under `fine`.

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,106 @@
+  make_option(
+    opt_str = c("--ref_file"),
+    type = "character",
+    help = "path to rds file with input sce object to be processed"
```
```suggestion
    help = "path to rds file with input sce object to build a SingleR model from"
```

sjspielman replied:
We might expand this to state that it must have `label.fine` and `label.main`.
I'd also add some quick checks that those columns exist before training the model.

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,106 @@
+  make_option(
+    opt_str = c("--output_file"),
+    type = "character",
+    help = "path to output rds file to store processed sce object. Must end in .rds"
```
Up to you

```suggestion
    help = "path to output rds file to store processed sce object. Must have `.rds` extension."
```

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,106 @@
+  make_option(
+    opt_str = c("--fry_tx2gene"),
+    type = "character",
+    help = "path to "
```
where?

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,106 @@
+
+# This script is used to build and train a model to be used for SingleR. 
+# The expected input is a SummarizedExperiment containing a reference dataset and a
+# 3-column, transcript to gene tsv file (as used with Alevin-fry), to provide the expected
```
Is this an internal `scpca-nf` format, or `Alevin-fry` format? My question really is whether it's worth explicitly stating what those 3 columns are.

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,106 @@
+                           col_names = c("transcript", "gene", "transcript_type"))
+
+# select genes to use for model restriction 
+gene_ids <- tx2gene |>
+  dplyr::select(gene) |>
+  unique()
```
This is a pretty simple "pipeline" so could be base R, but I don't feel strongly

```suggestion
gene_ids <- unique(tx2gene$gene)
```

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,106 @@
+fine_model <- SingleR::trainSingleR(
+  ref_data,
+  labels = ref_data$label.fine,
+  genes = "de", 
```
Hm, this could be a function argument for other use cases?

allyhawkins replied:
My understanding based on the documentation is that this is the only option we can use with our specified input. Otherwise we would need to provide a list of genes as input, which we are not going to be doing here. 
https://rdrr.io/github/LTLA/SingleR/man/trainSingleR.html

sjspielman reviewed (commented):
Looks about right to me, left some suggestions!

allyhawkins commented:
Thanks for taking a look at this, I implemented your comments and added a check for the labels being in the `colData` of the ref data that we are reading in. Also apologies because I hadn't fully filled out the help messages the first time around, but I fixed that this time. This should be ready for another round of review.

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,115 @@
+  make_option(
+    opt_str = c("--output_file"),
+    type = "character",
+    help = "path to output rds file to store trained model for reference dataset"
```
```suggestion
    help = "path to output rds file to store trained model for reference dataset."
```

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,115 @@
+    opt_str = c("-t", "--threads"),
+    type = "integer",
+    default = 1,
+    help = "Number of multiprocessing threads to use"
```
```suggestion
    help = "Number of multiprocessing threads to use."
```

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,115 @@
+
+opt <- parse_args(OptionParser(option_list = option_list))
+
+opt$ref_file <- "/Users/allyhawkins/Documents/ALSF/git_repos/ScPCA-admin/sample_info/celltype_annotation/references/celldex-blueprint_encode.rds"
```
Ha, I'm no stranger to this kind of local development bug, glad to see I'm not alone 😉

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,115 @@
+
+# check that input files exist
+if(!file.exists(opt$ref_file)){
+  stop("Missing input file with cell type reference.")
```
```suggestion
  stop("Missing input file with cell type reference (`ref_file`).")
```

sjspielman commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,115 @@
+                           col_names = c("transcript", "gene", "transcript_type"))
+
+# select genes to use for model restriction
+gene_ids <- unique(tx2gene$gene)
```
Possibly overkill, but maybe we should check this !=0? Or warn if it's below a certain number?

sjspielman reviewed (approved):
Looks good to me!

jashapiro reviewed (commented):
This looks good to me, but as discussed in our meeting, I think we should add `label.ont` to the training here. Which makes three trainings, which passes my threshold for converting to `purrr` or the like...

So I would do something like:
```
label_cols <- c("label.main", "label.fine", "label.ont")
names(label_cols) <- label_cols

models <- label_cols |>
  purrr::map(\(col) dplyr::pull(ref_data, var = col) |>
  purrr::map(
  \(labels) {SingleR::trainSingleR(
    ref_data,
    labels = labels,
    genes = "de",
    # only use genes found in index
    restrict = gene_ids,
    BPPARAM = bp_param
  })
```
 

A nice side benefit is that if we do need to add support for other label columns from other data sources, we only have to change the input label column list. Well, for this section anyway.

allyhawkins commented:
I made a few updates to this to incorporate using the `label.ont` as we discussed yesterday and mentioned in https://github.com/AlexsLemonade/scpca-nf/pull/280#pullrequestreview-1308002408. It would be great if reviewers could take one more look since I made those changes.

jashapiro commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,108 @@
+label_cols <- c("label.main", "label.fine", "label.ont")
+names(label_cols) <- label_cols
+
+models <- label_cols |>
+ purrr::map(\(labels) {SingleR::trainSingleR(
+                   ref_data,
+                   labels = ref_data[[labels]],
+                   genes = "de",
+                   # only use genes found in index
+                   restrict = gene_ids,
+                   BPPARAM = bp_param
+                 )})
```
Mostly formatting, but also a variable renaming.
```suggestion
models <- label_cols |>
 purrr::map(\(label_col) {
   SingleR::trainSingleR(
     ref_data,
     labels = ref_data[[label_col]],
     genes = "de",
     # only use genes found in index
     restrict = gene_ids,
     BPPARAM = bp_param    
   )})
```

jashapiro commented on bin/train_SingleR.R:
```diff
@@ -0,0 +1,108 @@
+models <- label_cols |>
+ purrr::map(\(labels) {SingleR::trainSingleR(
+                   ref_data,
+                   labels = ref_data[[labels]],
```
I'm assuming this line is correct, though it is quite unintuitive! 

I might substitute in `colData(ref_data)[, label_col]` to be more explicit?

allyhawkins replied:
Yes I tested it to make sure it works, but I can change it to be more explicit.

jashapiro reviewed (approved):
This looks good; I only had a couple of minor thoughts.
